### PR TITLE
[FTheoryTools] Rewrite _kodaira_type to work with any locus (Part 2)

### PR DESCRIPTION
**Pull request by HereAround**
cc @apturner (cf. https://github.com/oscar-system/Oscar.jl/pull/3424)

### CODE REVIEW

File: experimental/FTheoryTools/src/auxiliary.jl [hunk 1]
```diff
@@ -176,24 +176,59 @@ function _kodaira_type(id::MPolyIdeal{T}, f::T, g::T, d::T, ords::Tuple{Int64, I
...
     
     if f_ord == 0 && g_ord == 0
-      monodromy_poly = _psi^2 + divexact(evaluate(9 * poly_g, [locus], [0]), evaluate(2 * poly_f, [locus], [0]))
+      g_quotient = divrem(9 * poly_g, locus)[2]
+      f_quotient = divrem(2 * poly_f, locus)[2]
+      quotient_val = div(g_quotient, f_quotient)
+
+      # quotient_gens = gens(groebner_basis(quotient(ideal([9 * poly_g]), ideal([2 * poly_f])) + locus_id))
+      # deleteat!(quotient_gens, findall(g -> g == locus, quotient_gens))
+      # quotient_val = quotient_gens[1]
```
Comment (HereAround) [inline experimental/FTheoryTools/src/auxiliary.jl:188]: Is this still needed? If not, maybe just remove it?

File: experimental/FTheoryTools/src/auxiliary.jl [hunk 2]
```diff
@@ -176,24 +176,59 @@ function _kodaira_type(id::MPolyIdeal{T}, f::T, g::T, d::T, ords::Tuple{Int64, I
...
+
+      monodromy_poly = _psi^2 + quotient_val
       kod_type = _string_from_factor_count(monodromy_poly, ["Non-split I_$d_ord", "Split I_$d_ord"])
     elseif d_ord == 4 && g_ord == 2 && f_ord >= 2
-      monodromy_poly = _psi^2 - evaluate(divexact(poly_g, locus^2), [locus], [0])
+      g_quotient = divrem(div(poly_g, locus^2), locus)[2]
+
+      # quotient_gens = gens(groebner_basis(ideal([div(poly_g, locus^2)]) + locus_id))
+      # deleteat!(quotient_gens, findall(g -> g == locus, quotient_gens))
+      # quotient_val = quotient_gens[1]
```
Comment (HereAround) [inline experimental/FTheoryTools/src/auxiliary.jl:197]: Likewise.

File: experimental/FTheoryTools/src/auxiliary.jl [hunk 3]
```diff
@@ -176,24 +176,59 @@ function _kodaira_type(id::MPolyIdeal{T}, f::T, g::T, d::T, ords::Tuple{Int64, I
...
+
+      monodromy_poly = _psi^2 - g_quotient
       kod_type = _string_from_factor_count(monodromy_poly, ["Non-split IV", "Split IV"])
     elseif d_ord == 6 && f_ord >= 2 && g_ord >= 3
-      monodromy_poly =  _psi^3 + _psi * evaluate(divexact(poly_f, locus^2), [locus], [0]) + evaluate(divexact(poly_g, locus^3), [locus], [0])
+      f_quotient = divrem(div(poly_f, locus^2), locus)[2]
+      g_quotient = divrem(div(poly_g, locus^3), locus)[2]
+
+      # quotient_gens = gens(groebner_basis(ideal([_psi^3 + _psi * div(poly_f, locus^2) + div(poly_g, locus^3)]) + locus_id))
+      # deleteat!(quotient_gens, findall(g -> g == locus, quotient_gens))
```
Comment (HereAround) [inline experimental/FTheoryTools/src/auxiliary.jl:206]: As above.

File: experimental/FTheoryTools/src/auxiliary.jl [hunk 4]
```diff
@@ -176,24 +176,59 @@ function _kodaira_type(id::MPolyIdeal{T}, f::T, g::T, d::T, ords::Tuple{Int64, I
...
+      d_quotient = div(poly_d, locus^d_ord)
+      f_quotient = div(2 * poly_f, locus^2)
+      g_quotient = div(9 * poly_g, locus^3)
+      num_quotient = divrem(d_quotient * f_quotient^(2 + d_ord % 2), locus)[2]
+      den_quotient = divrem(4 * g_quotient^(2 + d_ord % 2), locus)[2]
+      quotient_val = div(num_quotient, den_quotient)
+
+      # quotient_4_gens = gens(groebner_basis(quotient(ideal([quotient_1 * quotient_2]) + locus_id, ideal([quotient_3]))))
+      # deleteat!(quotient_4_gens, findall(g -> g == locus, quotient_4_gens))
+      # quotient_4_val = quotient_4_gens[1]
```
Comment (HereAround) [inline experimental/FTheoryTools/src/auxiliary.jl:220]: Same here too.

File: experimental/FTheoryTools/src/auxiliary.jl [hunk 5]
```diff
@@ -176,24 +176,59 @@ function _kodaira_type(id::MPolyIdeal{T}, f::T, g::T, d::T, ords::Tuple{Int64, I
...
+
+      monodromy_poly = _psi^2 + quotient_val
       kod_type = _string_from_factor_count(monodromy_poly, ["Non-split I^*_$(d_ord - 6)", "Split I^*_$(d_ord - 6)"])
     elseif d_ord == 8 && g_ord == 4 && f_ord >= 3
-      monodromy_poly = _psi^2 - evaluate(divexact(poly_g, locus^4), [locus], [0])
+      g_quotient = divrem(div(poly_g, locus^4), locus)[2]
+
+      # quotient_gens = gens(groebner_basis(ideal([div(poly_g, locus^4)]) + locus_id))
+      # deleteat!(quotient_gens, findall(g -> g == locus, quotient_gens))
+      # quotient_val = quotient_gens[1]
```
Comment (HereAround) [inline experimental/FTheoryTools/src/auxiliary.jl:229]: And here.

File: experimental/FTheoryTools/test/tate_models.jl [hunk 6]
```diff
@@ -233,7 +233,7 @@ end
   irr_i5_s = irrelevant_ideal(tas);
   sri_i5_s = stanley_reisner_ideal(tas);
   lin_i5_s = ideal_of_linear_relations(tas);
-  id_fin,  = _blowup_global_sequence(id_i5_s, [[7, 8, 6], [2, 3, 1], [3, 4], [2, 4]], irr_i5_s, sri_i5_s, lin_i5_s)
+  id_fin,  = _blowup_global_sequence(id_i5_s, [[8, 9, 6], [2, 3, 1], [3, 4], [2, 4]], irr_i5_s, sri_i5_s, lin_i5_s)
```
Comment (HereAround) [inline experimental/FTheoryTools/test/tate_models.jl:236]: Why is this change needed? I was under the impression that this change would only affect the behavior "under the hood", but this change indicates differently. Or was there a mistake in the tests to begin with?

File: experimental/FTheoryTools/src/auxiliary.jl [hunk 7]
```diff
@@ -176,24 +176,59 @@ function _kodaira_type(id::MPolyIdeal{T}, f::T, g::T, d::T, ords::Tuple{Int64, I
...
+      f_quotient = divrem(div(poly_f, locus^2), locus)[2]
+      g_quotient = divrem(div(poly_g, locus^3), locus)[2]
+
+      # quotient_gens = gens(groebner_basis(ideal([_psi^3 + _psi * div(poly_f, locus^2) + div(poly_g, locus^3)]) + locus_id))
+      # deleteat!(quotient_gens, findall(g -> g == locus, quotient_gens))
+      
+      monodromy_poly = _psi^3 + _psi * f_quotient + g_quotient
       kod_type = _string_from_factor_count(monodromy_poly, ["Non-split I^*_0", "Semi-split I^*_0", "Split I^*_0"])
-    elseif f_ord == 2 && g_ord == 3 && d_ord >= 7 && d_ord % 2 == 1
-      monodromy_poly = _psi^2 + divexact(evaluate(divexact(poly_d, locus^d_ord) * divexact(2 * poly_f, locus^2)^3, [locus], [0]), 4 * evaluate(divexact(9 * poly_g, locus^3), [locus], [0])^3)
```
Comment (HereAround) [inline experimental/FTheoryTools/src/auxiliary.jl:190]: With an eye towards migrating eventually into `source` and satisfying all `OSCAR` code standards - this line is pretty hard to read because it is so long. And having several combined calls to `divexact` and `evaluate` does not exactly make the reading any easier (at least for me).

Suggestions:
Since the computation for the monodromy polynomial happens multiple times, maybe we could add an auxiliary function that completes this task? And this auxiliary function has strictly more than one line of code, maybe also a few comments on what exactly it does?

**Comment by HereAround** [comment]
Our to-do list (https://github.com/oscar-system/Oscar.jl/issues/2699) just reminded me of the following:

"Note: Quite a few places in the code use this assumption, and should be modified once the limitation is removed."

This applies to the literature models: When we build a model (presumably over a concrete base, but also needs to be checked about the arbitrary base) by specifying a line bundle as "gauge divisor", then it seems natural that this method picks the generic section of said line bundle and then declares the vanishing locus of this generic section as the gauge divisor. Currently, this does not happen.

This change should be easy and - as I see it - should happen in this PR.

I am not sure if there are other places that make a similar assumption - we should check.

File: experimental/FTheoryTools/src/auxiliary.jl (hunk 1 above)
Comment (apturner) [reply experimental/FTheoryTools/src/auxiliary.jl:188]: Sorry, I intended to remove all of these commented blocks that were in place during testing, but apparently neglected to, I will remove them

File: experimental/FTheoryTools/test/tate_models.jl (hunk 6 above)
Comment (apturner) [reply experimental/FTheoryTools/test/tate_models.jl:236]: This change is needed because an additional coordinate was added to the base space for the tests, in order to make the gauge locus more complicated

File: experimental/FTheoryTools/src/auxiliary.jl (hunk 7 above)
Comment (apturner) [reply experimental/FTheoryTools/src/auxiliary.jl:190]: The line you're commenting on here is removed in this PR, and I think the current version of the function doesn't have any excessively long lines in it. The reason I didn't add a helper function is because the monodromy polynomial is different for every singularity type. The bulk of the `_kodaira_type` function is this case-by-case computation of the monodromy polynomial anyway.

I do, however, agree that more comments would be helpful, I will add some in!

**Comment by HereAround** [comment]
Ok. Back to https://github.com/oscar-system/Oscar.jl/pull/3424... sorry for the commotion.

File: experimental/FTheoryTools/src/auxiliary.jl (hunk 7 above)
Comment (HereAround) [reply experimental/FTheoryTools/src/auxiliary.jl:190]: Pardon. My bad.
